ocaml/typing/typeopt.ml

```diff
@@ -196,39 +198,79 @@ let value_kind_of_value_layout layout =
+
+     type t = int * M.t
+
+   If we're missing the cmi for [M], we can't verify the invariant that
+   [value_kind] is only called on values.  However, we still know the pair
+   itself is a value, so a sound thing to do is fall back and return [Pgenval]
+   for [t].
+
+   On the other hand, if we're asked to compute the value kind for [M.t]
+   directly and are missing the cmi for [M], we really do need to issue an error.
+   This is a bug in the typechecker, which should have checked that the type
+   in question has layout value.
+
+   The implementation of this fallback behavior is combined with the layouts
+   safety check at the top of [value_kind].  The safety check tests whether the
+   type has layout value.  If it see [void], [float64] or an [any] that doesn't
+   arise from a missing cmi, it issues the loud error.  If it sees [any] that
+   does arise from a missing cmi, it raises [Missing_cmi_fallback].
+
+   In places where we're computing value_kinds for a bunch of subcomponents of a
+   type, we catch [Missing_cmi_fallback] and just return [Pgenval] for the outer
+   type.  In the top-level [value_kind], we catch it and issue the loud error.
+
+   The old world view was that we would drop the safety check from [value_kind]
+   after a while, once we felt confident in the type checking (it isn't free,
+   after all).  But that's wrong - we will always need the check here to make
+   sure we're sound in the event of a missing cmi.  Even when we change the
+   build system to begin passing the cmis for all transitive dependencies to the
+   compiler, we shouldn't be unsound in the event the compiler is invoked
+   manually or the transitive deps logic has a bug.
+
+   (But, if we ever do find a way to get rid of the safety check: Note that the
+   it is currently doing some defaulting of sort variables, as in cases like:
 
-   let () =
-     match assert false  with
-     | _ -> assert false
+     let () =
+       match assert false  with
+       | _ -> assert false
 
    There is a sort variable for the scrutinee of the match in typedtree that is
-   still a sort variable after checking this.  And this is fine - we default
-   sorts that appear in interfaces, but other sorts may still appear in the
-   typedtree.  These represent places where the compiler can pick a sort, and
-   should pick whatever is most efficient (eventually void).
+   still a sort variable after checking this.  It's fine to default this to
+   anything - void would be ideal, but for now it gets value.  If the safety check
+   goes away, think about whether we should add defaulting elsewhere.)
+*)
+exception Missing_cmi_fallback
+
+let fallback_if_missing_cmi fallback f =
+  try f () with Missing_cmi_fallback -> fallback
 
-   When we eliminate the safety check here, we should think again about where
-   and how sort variables are defaulted.
+(* CR layouts v2.5: It will be possible for subcomponents of types like pairs
```

Comment:
As I know you're aware, these comments are the same. What I hadn't realized previously, though, is that we have the problem in 2.5, because of records-of-unboxed-floats.